src/detectors/EnvDetector.ts

```diff
@@ -0,0 +1,182 @@
+      }
+    } else {
+      diag.debug(
+        'EnvDetector failed: Environment variable "OTEL_RESOURCE_ATTRIBUTES" is missing.'
+      );
+    }
+
+    if (serviceName) {
+      attributes[SemanticResourceAttributes.SERVICE_NAME] = serviceName;
+    }
+
+    return new Resource(attributes);
+  }
+
+  /**
+   * Creates an attribute map from the OTEL_RESOURCE_ATTRIBUTES environment
+   * variable.
+   *
+   * OTEL_RESOURCE_ATTRIBUTES: A comma-separated list of attributes describing
+   * the source in more detail, e.g. “key1=val1,key2=val2”. Domain names and
+   * paths are accepted as attribute keys. Values may be quoted or unquoted in
+   * general. If a value contains whitespaces, =, or " characters, it must
+   * always be quoted.
+   *
+   * @param rawEnvAttributes The resource attributes as a comma-seperated list
+   * of key/value pairs.
+   * @returns The sanitized resource attributes.
+   */
+  private _parseResourceAttributes(
+    rawEnvAttributes?: string
+  ): ResourceAttributes {
+    if (!rawEnvAttributes) return {};
+
+    const attributes: ResourceAttributes = {};
+    const rawAttributes: string[] = rawEnvAttributes.split(
+      this._COMMA_SEPARATOR,
+      -1
+    );
+    for (const rawAttribute of rawAttributes) {
+      const keyValuePair: string[] = rawAttribute.split(
+        this._LABEL_KEY_VALUE_SPLITTER,
+        -1
+      );
+      if (keyValuePair.length !== 2) {
+        continue;
+      }
+      let [key, value] = keyValuePair;
+      // Leading and trailing whitespaces are trimmed.
+      key = key.trim();
+      value = value.trim().split('^"|"$').join('');
+      if (!this._isValidAndNotEmpty(key)) {
+        throw new Error(`Attribute key ${this._ERROR_MESSAGE_INVALID_CHARS}`);
+      }
+      if (!this._isValid(value)) {
+        throw new Error(`Attribute value ${this._ERROR_MESSAGE_INVALID_VALUE}`);
+      }
+      attributes[key] = value;
+    }
+    return attributes;
+  }
+
+  /**
+   * Determines whether the given String is a valid printable ASCII string with
+   * a length not exceed _MAX_LENGTH characters.
+   *
+   * @param str The String to be validated.
+   * @returns Whether the String is valid.
+   */
+  private _isValid(name: string): boolean {
+    return name.length <= this._MAX_LENGTH && this._isPrintableString(name);
+  }
+
+  private _isPrintableString(str: string): boolean {
+    for (let i = 0; i < str.length; i++) {
+      const ch: string = str.charAt(i);
+      if (ch <= ' ' || ch >= '~') {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Determines whether the given String is a valid printable ASCII string with
+   * a length greater than 0 and not exceed _MAX_LENGTH characters.
+   *
+   * @param str The String to be validated.
+   * @returns Whether the String is valid and not empty.
+   */
+  private _isValidAndNotEmpty(str: string): boolean {
+    return str.length > 0 && this._isValid(str);
+  }
+}
+
+export const envDetector = new EnvDetector();
```

Comment:
If the detectors have no state, couldn't we just make them free functions?

Reply:
Why would we implement detectors differently than OTel has it?